# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.cpp
# Commit: f615134b7349
# Full Hash: f615134b73496943a0f2b088496dc6b2e3632a0b
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-29 15:31:45
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Implement VideoDecoder::Reset r=padenot
#   
#   This patch implements the reset method [1] for WebCodecs' VideoDecoder
#   interface. The implementation has not been completed yet. It will be
#   completed by the following patches.
# ==============================================================================

diff -r 4e7051264cb3 -r f615134b7349 dom/media/webcodecs/VideoDecoder.cpp
--- a/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 04:15:01 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 04:15:01 2023 +0000
@@ -157,7 +157,8 @@
                            RefPtr<VideoFrameOutputCallback>&& aOutputCallback)
     : DOMEventTargetHelper(aParent),
       mErrorCallback(std::move(aErrorCallback)),
-      mOutputCallback(std::move(aOutputCallback)) {
+      mOutputCallback(std::move(aOutputCallback)),
+      mState(CodecState::Unconfigured) {
   MOZ_ASSERT(mErrorCallback);
   MOZ_ASSERT(mOutputCallback);
   LOG("VideoDecoder %p ctor", this);
@@ -188,7 +189,8 @@
       RefPtr<VideoFrameOutputCallback>(aInit.mOutput));
 }
 
-CodecState VideoDecoder::State() const { return CodecState::EndGuard_; }
+// https://w3c.github.io/webcodecs/#dom-videodecoder-state
+CodecState VideoDecoder::State() const { return mState; }
 
 uint32_t VideoDecoder::DecodeQueueSize() const { return 0; }
 
@@ -212,8 +214,15 @@
   return nullptr;
 }
 
+// https://w3c.github.io/webcodecs/#dom-videodecoder-reset
 void VideoDecoder::Reset(ErrorResult& aRv) {
-  aRv.Throw(NS_ERROR_DOM_NOT_SUPPORTED_ERR);
+  AssertIsOnOwningThread();
+
+  LOG("VideoDecoder %p, Reset", this);
+
+  if (auto r = Reset(NS_ERROR_DOM_ABORT_ERR); r.isErr()) {
+    aRv.Throw(r.unwrapErr());
+  }
 }
 
 void VideoDecoder::Close(ErrorResult& aRv) {
@@ -264,6 +273,19 @@
   return p.forget();
 }
 
+// https://w3c.github.io/webcodecs/#reset-videodecoder
+Result<Ok, nsresult> VideoDecoder::Reset(const nsresult& aResult) {
+  AssertIsOnOwningThread();
+
+  if (mState == CodecState::Closed) {
+    return Err(NS_ERROR_DOM_INVALID_STATE_ERR);
+  }
+
+  mState = CodecState::Unconfigured;
+
+  return Ok();
+}
+
 #undef LOG
 #undef LOGE
 #undef LOG_INTERNAL